File: ms/services/product-service/product/src/main/java/chamika/product/controller/ProductController.java
package chamika.product.controller;

import chamika.product.dto.product.ProductApprovalStatusRequest;
import chamika.product.dto.product.ProductCreateReqBody;
import chamika.product.dto.product.ProductResponseBody;
import chamika.product.service.product.ProductService;
import chamika.product.shared.PageResponse;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

@RestController
@RequestMapping("/api/v1/product")
@RequiredArgsConstructor
public class ProductController {



    private final ProductService productService;


    @PostMapping
    public ResponseEntity<ProductResponseBody> createProduct(
            @RequestPart("product") @Valid ProductCreateReqBody reqBody,
            @RequestPart("images") List<MultipartFile> images
    ) {

        List<String> imageUrls = productService.uploadProductImages(images);

        ProductResponseBody product = productService.createProduct(reqBody, imageUrls);
        return ResponseEntity.status(HttpStatus.CREATED).body(product);
    }

    @GetMapping
    @ResponseStatus(HttpStatus.OK)
    public ResponseEntity<PageResponse<ProductResponseBody>> getAllProducts(
            @RequestParam(required = false) String query,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size,
            @RequestParam(defaultValue = "id") String sortBy,
            @RequestParam(defaultValue = "asc") String sortDir) {

        return ResponseEntity.ok(productService.getAllProducts(page, size, sortBy, sortDir, query));

    }


    @PutMapping("/{id}")
    @ResponseStatus(HttpStatus.OK)
    public ResponseEntity<ProductResponseBody> updateProduct(
            @PathVariable Long id,
            @RequestPart("product") @Valid ProductCreateReqBody reqBody
    ) {
        ProductResponseBody updatedProduct = productService.updateProduct(id, reqBody);
        return ResponseEntity.ok(updatedProduct);
    }


    @GetMapping("/{id}")
    @ResponseStatus(HttpStatus.OK)
    public ResponseEntity<ProductResponseBody> getProductById(@PathVariable Long id) {
        return ResponseEntity.ok(productService.getProductById(id));
    }


    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public ResponseEntity<Void> deleteProduct(@PathVariable Long id) {
        productService.deleteProduct(id);
        return ResponseEntity.noContent().build();
    }


   @PatchMapping("/{id}/approve")
    @ResponseStatus(HttpStatus.OK)
    public ResponseEntity<ProductResponseBody> approveProduct(
            @PathVariable Long id,
            @RequestBody @Valid ProductApprovalStatusRequest reqBody
   ) {
        return ResponseEntity.ok(productService.approveProduct(id, reqBody));
    }



}

File: backend_for_frontend/src/controllers/product.controller.ts
import { Request, Response } from "express";
import { ProductService } from "../services/product.service";
import { asyncHandler } from "../utils/error.handler";

export class ProductController {
  constructor(private readonly productService: ProductService) {}

  createProduct = asyncHandler(async (req: Request, res: Response) => {
    const product = await this.productService.createProduct(
      req.body.product,
      req.files as Express.Multer.File[]
    );
    res.status(201).json(product);
  });

  getAllProducts = asyncHandler(async (req: Request, res: Response) => {
    const {
      page = 0,
      size = 10,
      sortBy = "id",
      sortDir = "asc",
      query,
    } = req.query;

    const products = await this.productService.getAllProducts(
      Number(page),
      Number(size),
      String(sortBy),
      String(sortDir),
      query as string | undefined
    );
    res.json(products);
  });

  updateProduct = asyncHandler(async (req: Request, res: Response) => {
    const product = await this.productService.updateProduct(
      Number(req.params.id),
      req.body.product
    );
    res.json(product);
  });

  getProductById = asyncHandler(async (req: Request, res: Response) => {
    const product = await this.productService.getProductById(
      Number(req.params.id)
    );
    res.json(product);
  });

  deleteProduct = asyncHandler(async (req: Request, res: Response) => {
    await this.productService.deleteProduct(Number(req.params.id));
    res.status(204).send();
  });

  approveProduct = asyncHandler(async (req: Request, res: Response) => {
    const product = await this.productService.approveProduct(
      Number(req.params.id),
      req.body
    );
    res.json(product);
  });
}
